examples/cluster/plot_kmeans_silhouette_analysis.py

```diff
@@ -0,0 +1,139 @@
+
+from sklearn.datasets import make_blobs
+from sklearn.cluster import KMeans
+from sklearn.metrics import silhouette_samples, silhouette_score
+
+import matplotlib.pyplot as plt
+import matplotlib.cm as cm
+import numpy as np
+
+print(__doc__)
+
+# Generating the sample data from make_blobs
+# This particular setting has one distict cluster and 3 clusters placed close
+# together.
+X, y = make_blobs(n_samples=500,
+                  n_features=2,
+                  centers=4,
+                  cluster_std=1.0,
+                  center_box=(-10.0, 10.0),
+                  shuffle=True,
+                  random_state=1)  # For reproducibility
+
+range_n_clusters = [2, 4, 6]
+
+for n_clusters in range_n_clusters:
+    # Create a subplot with 1 row and 2 columns
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    fig.set_size_inches(18, 7)
+
+    # The 1st subplot is the silhouette plot
+    # The silhouette coefficient can range from -1, 1 but in this example all
+    # lie within [-0.1, 1]
+    ax1.set_xlim([-0.1, 1])
+    # The n_clusters*10 are the additional samples to demarcate the space
+    # between silhouette plots of individual clusters.
+    ax1.set_ylim([0, len(X) + n_clusters * 10])
+
+    # Initialize the clusterer with n_clusters value and a random generator
+    # seed of 10 for reproducibility.
+    clusterer = KMeans(n_clusters=n_clusters, random_state=10)
+    cluster_labels = clusterer.fit_predict(X)
+
+    # The silhouette_score gives the average value for all the samples.
+    # This gives a perspective into the density and separation of the formed
+    # clusters
+    silhouette_avg = silhouette_score(X, cluster_labels)
+    print("For n_clusters =", n_clusters,
+          "The average silhouette_score is :", silhouette_avg)
+
+    # Compute the silhouette scores for each sample
+    sample_silhouette_values = silhouette_samples(X, cluster_labels)
+
+    # This will hold the silhouette coefficient of all the clusters separated
+    # by 0 samples.
+    sorted_clustered_sample_silhouette_values = []
+
+    for i in np.unique(cluster_labels):
+        ith_cluster_silhouette_values = \
+            sample_silhouette_values[cluster_labels == i]
+
+        # Add the ith_cluster_silhouette_values after sorting them
+        ith_cluster_silhouette_values.sort()
+
+        # The introduced 0 samples are to differentiate clearly between the
+        # different clusters
+        sorted_clustered_sample_silhouette_values += \
+            ith_cluster_silhouette_values.tolist() + [0] * 10
+
+    x_values = np.array(sorted_clustered_sample_silhouette_values)
+    y_range = np.arange(len(X) + 10 * n_clusters)
+
+    # Computing custom label coordinates for labeling the clusters
+    # Plot the silhouette with the corresponding cluster color
+    offset = 0
+    for i in range(n_clusters):
+        size_cluster_i = sum(cluster_labels == i)
+
+        x = -0.05
+        # Label them at the middle
+        dy = size_cluster_i
```

Comment:
why do you join the lists and then disentangle them again? you could just plot as you compute using a single loop, right?